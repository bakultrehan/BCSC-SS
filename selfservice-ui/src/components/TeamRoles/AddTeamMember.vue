/** * Add team member of app */

<template>
  <v-container>
    <v-row class="text-left">
      <v-col v-if="userDetails">
        <v-card class="v-form ma-3" flat>
          <v-form ref="form" v-model="form">
            <v-card class="v-form px-6 ma-3">
              <v-alert type="error" dense outlined class="text-left" v-if="memberErrorStatus">
                <div>{{ $t('addTeamMember.errorTitle') }}</div>
                <ul>
                  <li
                    v-for="(errors, idx) in errorList"
                    :key="idx"
                    v-html="$t(`addTeamMember.${errors}`)"
                  ></li>
                </ul>
              </v-alert>

              <v-row>
                <v-col cols="12">
                  <div class="display-1">
                    {{
                    $t(
                    !editMode
                    ? 'addTeamMember.pagetitle'
                    : 'addTeamMember.pagetitleUpdate'
                    )
                    }}
                  </div>
                </v-col>
                <v-col cols="12" sm="6">
                  <v-card-subtitle class="headline bc-padding-left-0">
                    {{
                    $t('addTeamMember.contactInfo')
                    }}
                  </v-card-subtitle>
                  <Input
                    v-model="userDetails.firstName"
                    :label="$t('addTeamMember.labelFirstName')"
                    :rules="[
                      rules.required,
                      rules.length(1),
                      rules.maxLength(250)
                    ]"
                    type="text"
                    :disabled="disabled()"
                  />
                  <Input
                    v-model="userDetails.lastName"
                    :label="$t('addTeamMember.labelLastName')"
                    :rules="[
                      rules.required,
                      rules.length(1),
                      rules.maxLength(250)
                    ]"
                    type="text"
                    :disabled="disabled()"
                  />
                  <Input
                    v-model="userDetails.email"
                    :label="$t('addTeamMember.labelWorkEmail')"
                    :rules="[rules.required, rules.email, rules.maxLength(250)]"
                    type="text"
                    :disabled="disabled()"
                  />

                  <Input
                    v-model="userDetails.phone"
                    :label="$t('addTeamMember.labelPhone')"
                    type="text"
                    :disabled="disabled()"
                    :optional="true"
                  />
                </v-col>
                <v-col>
                  <v-divider class="mx-4" :inset="inset" vertical></v-divider>
                </v-col>
                <v-col cols="12" sm="5">
                  <v-card-subtitle class="headline bc-padding-left-0">
                    {{
                    $t('addTeamMember.roleTitle')
                    }}
                  </v-card-subtitle>

                  <v-radio-group v-model="userDetails.role" :mandatory="false">
                    <v-radio
                      :label="$t(`addTeamMember.labelRole${rolesList[1]}`)"
                      :value="projectRoles.developer"
                      class="my-2"
                    ></v-radio>
                    <v-radio
                      :label="$t(`addTeamMember.labelRole${rolesList[2]}`)"
                      :value="projectRoles.manager"
                      class="my-2"
                    ></v-radio>
                    <v-radio
                      :label="$t(`addTeamMember.labelRole${rolesList[3]}`)"
                      :value="projectRoles.cto"
                      class="my-2"
                    ></v-radio>
                  </v-radio-group>

                  <v-card-actions class="btn-bottom">
                    <v-spacer></v-spacer>
                    <Button
                      @click="$emit('toggleAddMember', false)"
                      aria-label="Back Button"
                      secondary
                    >{{ $t('addTeamMember.btnCancel') }}</Button>
                    <Button
                      :disabled="!form"
                      class="white--text submit-package ml-6"
                      @click="submitTeamMember"
                    >{{ $t('addTeamMember.btnSumbmit') }}</Button>
                  </v-card-actions>
                </v-col>
              </v-row>
            </v-card>
          </v-form>
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import { Component, Vue, Prop, Watch } from 'vue-property-decorator';
import { Getter, namespace, Action } from 'vuex-class';
// import { ProjectUserModel, ProjectInfoModel } from '@/models/ProjectInfoModel';
import Input from '@/Atomic/Input/Input.vue';
import Button from '@/Atomic/Button/Button.vue';

import validationRules from '@/config/validationRules';
import { projectRoles, projectRolesList } from '@/constants/enums';
import { TeamRoleModel } from '@/models/TeamRoleModel';
import { memberDetails } from '@/store/modules/TeamRoles/defaults';

const TeamRolesModule = namespace('TeamRolesModule');
const KeyCloakModule = namespace('KeyCloakModule');

@Component({
  components: {
    Input,
    Button
  }
})
export default class AddTeamMember extends Vue {
  @Prop({ default: 0 })
  public id!: number;
  @Prop()
  public toggleAddMember: any;
  @Prop({ default: 0 })
  public memberId!: number;

  @TeamRolesModule.Getter('getTeamList')
  public teamList!: any;
  @TeamRolesModule.Action('addTeamMember')
  public addTeamMember!: any;
  @TeamRolesModule.Action('updateTeamMember')
  public updateTeamMember!: any;
  @TeamRolesModule.Action('getTeamMember')
  public getTeamMember!: any;
  @TeamRolesModule.Getter('memberErrorStatus')
  public memberErrorStatus!: any;
  @TeamRolesModule.Getter('getMemberErrorList')
  public getMemberErrorList!: any;
  @TeamRolesModule.Getter('getMemberDetails')
  public memberDetails!: any;
  @TeamRolesModule.Getter('memberSucessStatus')
  public memberSucessStatus!: any;
  @TeamRolesModule.Action('clearMemberData')
  public clearMemberData!: any;

  public errorList: any = {};

  public userDetails: TeamRoleModel;

  @KeyCloakModule.Getter('isAdmin')
  private isAdmin!: any;

  private rules: any = validationRules;
  private projectRoles: any = projectRoles;
  private selectedRole: any = 1;
  private rolesList: any = projectRolesList;
  private form: boolean = false;
  private editMode: boolean = false;

  public constructor() {
    super();
    this.userDetails = memberDetails;
  }

  @Watch('getMemberErrorList')
  private ongetMemberErrorListChanged(val: any) {
    this.errorList = val;
  }

  @Watch('memberDetails')
  private ongetmemberDetailsChanged(val: any) {
    this.userDetails = val;
  }

  @Watch('memberId')
  private onmemberIdChanged(val: any) {
    this.memberId = val;
    this.getMemberDetails();
  }
  @Watch('memberSucessStatus')
  private onmemberSucessStatusChanged(val: boolean) {
    if (val) {
      this.$emit('toggleAddMember', false);
    }
  }

  private submitTeamMember() {
    if (this.editMode) {
      this.updateTeamMember({
        userDetails: this.userDetails,
        projectId: this.id,
        memberId: this.memberId
      });
    } else {
      this.addTeamMember({
        userDetails: this.userDetails,
        projectId: this.id
      });
    }
  }
  private getMemberDetails() {
    if (this.memberId !== 0) {
      this.editMode = true;
      this.getTeamMember({ memberId: this.memberId, projectId: this.id });
    } else {
      this.editMode = false;
      this.clearMemberData();
    }
  }
  private disabled() {
    if (this.editMode && !this.isAdmin) {
      return true;
    }
    return false;
  }

  private mounted() {
    this.getMemberDetails();
  }
}
</script>

<style lang="scss" scoped>
@import './../../assets/styles/theme.scss';
.v-text-field input {
  padding: 3px 0 3px 0;
}
.btn-bottom {
  bottom: 32px;
  position: absolute;
}
</style>
