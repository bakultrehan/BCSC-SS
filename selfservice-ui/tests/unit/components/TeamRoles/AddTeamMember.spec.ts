import { shallowMount } from '@vue/test-utils';
import AddTeamMember from '@/components/TeamRoles/AddTeamMember.vue';
import Vuetify from 'vuetify';
import Vuex from 'vuex';
let vuetify: any;
vuetify = new Vuetify();

describe('AddTeamMember.vue', () => {
  const store = new Vuex.Store({
    modules: {
      TeamRolesModule: {
        namespaced: true,
        state: {},
        getters: {
          successStatus: jest.fn(),
          errorStatus: jest.fn(),
          isLoading: jest.fn(),

          getMemberDetails: jest.fn(() => {
            return {};
          }),
          getMemberErrorList: jest.fn(),
          memberSucessStatus: jest.fn(),
          memberErrorStatus: jest.fn(),
        },
        actions: {
          clearMemberData: jest.fn(),
          getTeamMember: jest.fn(),
        },
      },
      KeyCloakModule: {
        namespaced: true,
        state: {},
        getters: {
          isAdmin: jest.fn(() => {
            return true;
          }),
        },
        actions: { setLogout: jest.fn() },
      },
    },
  });

  const mountFunction = (options: any) => {
    return shallowMount(AddTeamMember, {
      store,
      vuetify,
      mocks: { $t: jest.fn(() => {}) }, // tslint:disable-line
      ...options,
      stubs: {
        VForm: {
          render: () => jest.fn(),
          methods: {
            resetValidation: () => jest.fn(),
          },
        },
      },
    });
  };

  it('renders props when passed with gettors', () => {
    const AddTeamMemberPage = mountFunction({});
    expect(AddTeamMemberPage.element).toMatchSnapshot();
  });
  it('renders props when passed with gettors with member id', () => {
    const AddTeamMemberPage = mountFunction({
      propsData: {
        memberId: 2,
      },
    });
    expect(AddTeamMemberPage.element).toMatchSnapshot();
  });
});
