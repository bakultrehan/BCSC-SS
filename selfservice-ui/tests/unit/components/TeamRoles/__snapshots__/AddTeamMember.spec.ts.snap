// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTeamMember.vue renders props when passed with gettors 1`] = `
<v-container-stub
  tag="div"
>
  <v-row-stub
    class="text-left"
    tag="div"
  >
    <v-col-stub
      tag="div"
    >
      <v-card-stub
        class="v-form ma-3"
        flat="true"
        loaderheight="4"
        tag="div"
      >
        <v-form-stub>
          <v-card-stub
            class="v-form px-6 ma-3"
            loaderheight="4"
            tag="div"
          >
            <!---->
             
            <v-row-stub
              tag="div"
            >
              <v-col-stub
                cols="12"
                tag="div"
              >
                <div
                  class="display-1"
                >
                  
                  
                
                </div>
              </v-col-stub>
               
              <v-col-stub
                cols="12"
                sm="6"
                tag="div"
              >
                <v-card-subtitle-stub
                  class="headline bc-padding-left-0"
                >
                  
                  
                
                </v-card-subtitle-stub>
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => (v || '').length >= len || 'Invalid character length, required ' + len,v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => (v || '').length >= len || 'Invalid character length, required ' + len,v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => /^\\\\w+([.-]?\\\\w+)*@\\\\w+([.-]?\\\\w+)*(\\\\.\\\\w{2,10})+$/.test(v) || 'E-mail must be valid',v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  optional="true"
                  outlined="true"
                  type="text"
                  value=""
                />
              </v-col-stub>
               
              <v-col-stub
                tag="div"
              >
                <v-divider-stub
                  class="mx-4"
                  vertical="true"
                />
              </v-col-stub>
               
              <v-col-stub
                cols="12"
                sm="5"
                tag="div"
              >
                <v-card-subtitle-stub
                  class="headline bc-padding-left-0"
                >
                  
                  
                
                </v-card-subtitle-stub>
                 
                <v-radio-group-stub
                  activeclass="v-item--active"
                  backgroundcolor=""
                  column="true"
                  errorcount="1"
                  errormessages=""
                  height="auto"
                  messages=""
                  rules=""
                  successmessages=""
                  value="1"
                  valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
                >
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="1"
                  />
                   
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="2"
                  />
                   
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="3"
                  />
                </v-radio-group-stub>
                 
                <v-card-actions-stub
                  class="btn-bottom"
                >
                  <v-spacer-stub />
                   
                  <button-stub
                    aria-label="Back Button"
                    color=""
                    name="button"
                    secondary=""
                    type="button"
                  >
                    
                  </button-stub>
                   
                  <button-stub
                    class="white--text submit-package ml-6"
                    color=""
                    disabled="true"
                    name="button"
                    type="button"
                  >
                    
                  </button-stub>
                </v-card-actions-stub>
              </v-col-stub>
            </v-row-stub>
          </v-card-stub>
        </v-form-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;

exports[`AddTeamMember.vue renders props when passed with gettors with member id 1`] = `
<v-container-stub
  tag="div"
>
  <v-row-stub
    class="text-left"
    tag="div"
  >
    <v-col-stub
      tag="div"
    >
      <v-card-stub
        class="v-form ma-3"
        flat="true"
        loaderheight="4"
        tag="div"
      >
        <v-form-stub>
          <v-card-stub
            class="v-form px-6 ma-3"
            loaderheight="4"
            tag="div"
          >
            <!---->
             
            <v-row-stub
              tag="div"
            >
              <v-col-stub
                cols="12"
                tag="div"
              >
                <div
                  class="display-1"
                >
                  
                  
                
                </div>
              </v-col-stub>
               
              <v-col-stub
                cols="12"
                sm="6"
                tag="div"
              >
                <v-card-subtitle-stub
                  class="headline bc-padding-left-0"
                >
                  
                  
                
                </v-card-subtitle-stub>
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => (v || '').length >= len || 'Invalid character length, required ' + len,v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => (v || '').length >= len || 'Invalid character length, required ' + len,v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  outlined="true"
                  rules="v => !!v || 'This field is required',v => /^\\\\w+([.-]?\\\\w+)*@\\\\w+([.-]?\\\\w+)*(\\\\.\\\\w{2,10})+$/.test(v) || 'E-mail must be valid',v => (v || '').length <= len || 'Invalid character length, max length ' + len"
                  type="text"
                  value=""
                />
                 
                <input-stub
                  helptext=""
                  label=""
                  optional="true"
                  outlined="true"
                  type="text"
                  value=""
                />
              </v-col-stub>
               
              <v-col-stub
                tag="div"
              >
                <v-divider-stub
                  class="mx-4"
                  vertical="true"
                />
              </v-col-stub>
               
              <v-col-stub
                cols="12"
                sm="5"
                tag="div"
              >
                <v-card-subtitle-stub
                  class="headline bc-padding-left-0"
                >
                  
                  
                
                </v-card-subtitle-stub>
                 
                <v-radio-group-stub
                  activeclass="v-item--active"
                  backgroundcolor=""
                  column="true"
                  errorcount="1"
                  errormessages=""
                  height="auto"
                  messages=""
                  rules=""
                  successmessages=""
                  value="1"
                  valuecomparator="function deepEqual(a, b) {
  if (a === b) return true;

  if (a instanceof Date && b instanceof Date) {
    // If the values are Date, they were convert to timestamp with getTime and compare it
    if (a.getTime() !== b.getTime()) return false;
  }

  if (a !== Object(a) || b !== Object(b)) {
    // If the values aren't objects, they were already checked for equality
    return false;
  }

  var props = Object.keys(a);

  if (props.length !== Object.keys(b).length) {
    // Different number of props, don't bother to check
    return false;
  }

  return props.every(function (p) {
    return deepEqual(a[p], b[p]);
  });
}"
                >
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="1"
                  />
                   
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="2"
                  />
                   
                  <v-radio-stub
                    class="my-2"
                    officon="$radioOff"
                    onicon="$radioOn"
                    ripple="true"
                    value="3"
                  />
                </v-radio-group-stub>
                 
                <v-card-actions-stub
                  class="btn-bottom"
                >
                  <v-spacer-stub />
                   
                  <button-stub
                    aria-label="Back Button"
                    color=""
                    name="button"
                    secondary=""
                    type="button"
                  >
                    
                  </button-stub>
                   
                  <button-stub
                    class="white--text submit-package ml-6"
                    color=""
                    disabled="true"
                    name="button"
                    type="button"
                  >
                    
                  </button-stub>
                </v-card-actions-stub>
              </v-col-stub>
            </v-row-stub>
          </v-card-stub>
        </v-form-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
